ath79: add support for TP-Link WBS210 v2

citronalco opened the pull request:
Signed-off-by: Bernhard Geier <freifunk@geierb.de>

2.4GHz outdoor device with two detachable antennas. Technically similar to the TP-Link CPE220v2.

Successfully tested flashing from original firmware, flashing sysupgrade, LEDs, reset button, WiFi and both ethernet ports.

citronalco committed:
ath79: add support for TP-Link WBS210 v2
TP-Link WBS210 v2 is an outdoor wireless CPE for 2.4 GHz with
two Ethernet ports based on Atheros AR9344

The device is the same as TP-Link CPE220 v2, but with higher TX power (27 dBm
instead of 12 dBm) and two antenna connectors instead of built-in antennas.

Specifications:
 - SoC: Atheros AR9344
 - RAM: 64MB
 - Storage: 8 MB SPI NOR
 - Wireless: 2.4GHz 300 Mbps, 2x RP-SMA connector, 27 dBm TX power
 - Ethernet: 1x 10/100 Mbps with 24V POE IN, 1x 10/100 Mbps

Installation:
Flash factory image through stock firmware WEB UI
or through TFTP
To get to TFTP recovery just hold reset button while powering on for
around 4-5 seconds and release.
Rename factory image to recovery.bin
Stock TFTP server IP: 192.168.0.100
Stock device TFTP adress: 192.168.0.254

Signed-off-by: Bernhard Geier <freifunk@geierb.de>

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -168,6 +168,7 @@ tplink,archer-d50-v1)
 	ucidef_set_led_switch "wan_data" "WAN Data" "tp-link:white:internet" "switch0" "0x02" "" "tx rx"
 	ucidef_set_led_switch "wan_link" "WAN Link" "tp-link:white:wan" "switch0" "0x02" "" "link"
 	;;
+tplink,wbs210-v2|\
```
Please add the device identifiers so that alphabetic order is maintained (also in all cases below).

adrianschmutzler commented on target/linux/ath79/image/generic-tp-link.mk:
```diff
@@ -340,6 +340,18 @@ define Device/tplink_re450-v2
+  DEVICE_VARIANT := v2
+  DEVICE_PACKAGES := rssileds
+  TPLINK_BOARD_ID := WBS210V2
+  SUPPORTED_DEVICES += wbs210-v2
```
Drop this. No support in ar71xx.

citronalco replied:
What exactly?

adrianschmutzler replied:
Drop the SUPPORTED_DEVICES line.

adrianschmutzler commented:
Hi, please improve your commit title/message:
- Add ath79 prefix to commit title
- Add some description, device specs, flashing instructions to commit message

Maybe also have a look at https://openwrt.org/submitting-patches

Despite, have you checked the partition layout or just copied it from an existing device?

citronalco commented:
Fixed commit title, commit message, alphabetical order and removed the "SUPPORTED_DEVICES" line, hope it's better now.

I checked the partition layout.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/01_leds:
```diff
@@ -170,7 +170,8 @@ tplink,archer-d50-v1)
 tplink,cpe210-v1|\
 tplink,cpe220-v2|\
-tplink,cpe510-v1)
+tplink,cpe510-v1)\
```
Typo at the end of line.

adrianschmutzler commented on target/linux/ath79/base-files/etc/board.d/03_gpio_switches:
```diff
@@ -44,7 +44,8 @@ tplink,archer-c25-v1)
 tplink,cpe210-v1|\
 tplink,cpe220-v2|\
-tplink,cpe510-v1)
+tplink,cpe510-v1)\
```
Same typo as above.

adrianschmutzler commented:
Hi, thanks for your changes. Please squash both patches into one.

In your commit message:

> The device is very similar to TP-Link CPE220 v2

Either state what's similar and different or remove that statement, as it's not helping without additional information.

citronalco commented:
I really appreciate your help.
Better now?

adschm reviewed (approved):
Based on the information given that this is just another CPE clone, the code in this PR looks good to me.

psyborg55 commented:
Can you provide bootlog? It would be good to test multiple units in order to make sure wifi bug present with CPE210 v2/v3 does not happen with this one.

adrianschmutzler commented:
@psyborg55 Note that this is ar9344 (so, same as two-port CPE210 v1), while one-port CPE210 v2/v3 is qca9533.

There is no report about similar issues on ar9344 so far.

psyborg55 commented:
Still, show us kernel log after boot

citronalco commented:
Here's the output of dmesg:
```[    0.000000] Linux version 4.19.75 (geierb@fuckup4) (gcc version 7.4.0 (OpenWrt GCC 7.4.0 r11143-881ed09ee6)) #0 Sun Sep 29 11:51:28 2019
[    0.000000] bootconsole [early0] enabled
[    0.000000] CPU0 revision is: 0001974c (MIPS 74Kc)
[    0.000000] MIPS: machine is TP-Link WBS210 v2
[    0.000000] SoC: Atheros AR9344 rev 2
[    0.000000] Determined physical RAM map:
[    0.000000]  memory: 04000000 @ 00000000 (usable)
[    0.000000] Initrd not found or empty - disabling initrd
[    0.000000] Primary instruction cache 64kB, VIPT, 4-way, linesize 32 bytes.
[    0.000000] Primary data cache 32kB, 4-way, VIPT, cache aliases, linesize 32 bytes
[    0.000000] Zone ranges:
[    0.000000]   Normal   [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Movable zone start for each node
[    0.000000] Early memory node ranges
[    0.000000]   node   0: [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] Initmem setup node 0 [mem 0x0000000000000000-0x0000000003ffffff]
[    0.000000] On node 0 totalpages: 16384
[    0.000000]   Normal zone: 144 pages used for memmap
[    0.000000]   Normal zone: 0 pages reserved
[    0.000000]   Normal zone: 16384 pages, LIFO batch:3
[    0.000000] random: get_random_bytes called from start_kernel+0x98/0x4a8 with crng_init=0
[    0.000000] pcpu-alloc: s0 r0 d32768 u32768 alloc=1*32768
[    0.000000] pcpu-alloc: [0] 0 
[    0.000000] Built 1 zonelists, mobility grouping on.  Total pages: 16240
[    0.000000] Kernel command line: console=ttyS0,115200 rootfstype=squashfs,jffs2
[    0.000000] Dentry cache hash table entries: 8192 (order: 3, 32768 bytes)
[    0.000000] Inode-cache hash table entries: 4096 (order: 2, 16384 bytes)
[    0.000000] Writing ErrCtl register=00000000
[    0.000000] Readback ErrCtl register=00000000
[    0.000000] Memory: 57576K/65536K available (4272K kernel code, 179K rwdata, 1008K rodata, 1188K init, 207K bss, 7960K reserved, 0K cma-reserved)
[    0.000000] SLUB: HWalign=32, Order=0-3, MinObjects=0, CPUs=1, Nodes=1
[    0.000000] NR_IRQS: 51
[    0.000000] CPU clock: 560.000 MHz
[    0.000000] clocksource: MIPS: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 6825930166 ns
[    0.000010] sched_clock: 32 bits at 280MHz, resolution 3ns, wraps every 7669584382ns
[    0.009075] Calibrating delay loop... 278.93 BogoMIPS (lpj=1394688)
[    0.086306] pid_max: default: 32768 minimum: 301
[    0.091922] Mount-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.099626] Mountpoint-cache hash table entries: 1024 (order: 0, 4096 bytes)
[    0.114331] clocksource: jiffies: mask: 0xffffffff max_cycles: 0xffffffff, max_idle_ns: 19112604462750000 ns
[    0.125814] futex hash table entries: 256 (order: -1, 3072 bytes)
[    0.133067] pinctrl core: initialized pinctrl subsystem
[    0.140371] NET: Registered protocol family 16
[    0.152750] GPIO line 18 (tp-link:ext:lna0) hogged as output/high
[    0.159903] GPIO line 19 (tp-link:ext:lna1) hogged as output/high
[    0.196623] clocksource: Switched to clocksource MIPS
[    0.203988] NET: Registered protocol family 2
[    0.210066] tcp_listen_portaddr_hash hash table entries: 512 (order: 0, 4096 bytes)
[    0.219102] TCP established hash table entries: 1024 (order: 0, 4096 bytes)
[    0.227236] TCP bind hash table entries: 1024 (order: 0, 4096 bytes)
[    0.234630] TCP: Hash tables configured (established 1024 bind 1024)
[    0.242200] UDP hash table entries: 256 (order: 0, 4096 bytes)
[    0.249063] UDP-Lite hash table entries: 256 (order: 0, 4096 bytes)
[    0.256852] NET: Registered protocol family 1
[    0.261954] PCI: CLS 0 bytes, default 32
[    0.266300] Crashlog allocated RAM at address 0x3f00000
[    0.274204] workingset: timestamp_bits=14 max_order=14 bucket_order=0
[    0.290034] squashfs: version 4.0 (2009/01/31) Phillip Lougher
[    0.296872] jffs2: version 2.2 (NAND) (SUMMARY) (LZMA) (RTIME) (CMODE_PRIORITY) (c) 2001-2006 Red Hat, Inc.
[    0.324229] io scheduler noop registered
[    0.328862] io scheduler deadline registered (default)
[    0.337304] pinctrl-single 1804002c.pinmux: 544 pins, size 68
[    0.345160] Serial: 8250/16550 driver, 1 ports, IRQ sharing disabled
[    0.353465] console [ttyS0] disabled
[    0.357723] 18020000.uart: ttyS0 at MMIO 0x18020000 (irq = 9, base_baud = 2500000) is a 16550A
[    0.367765] console [ttyS0] enabled
[    0.375182] bootconsole [early0] disabled
[    0.401215] m25p80 spi0.0: s25fl064k (8192 Kbytes)
[    0.406147] 6 fixed-partitions partitions found on MTD device spi0.0
[    0.412634] Creating 6 MTD partitions on "spi0.0":
[    0.417532] 0x000000000000-0x000000020000 : "u-boot"
[    0.423620] 0x000000020000-0x000000030000 : "partition-table"
[    0.430567] 0x000000030000-0x000000040000 : "info"
[    0.436390] 0x000000040000-0x0000007c0000 : "firmware"
[    0.442644] 2 fixed-partitions partitions found on MTD device firmware
[    0.449349] Creating 2 MTD partitions on "firmware":
[    0.454398] 0x000000000000-0x000000200000 : "kernel"
[    0.460435] 0x000000200000-0x000000780000 : "rootfs"
[    0.466367] mtd: device 5 (rootfs) set to be root filesystem
[    0.477378] 1 squashfs-split partitions found on MTD device rootfs
[    0.483679] 0x000000400000-0x000000780000 : "rootfs_data"
[    0.490228] 0x0000007c0000-0x0000007f0000 : "config"
[    0.496219] 0x0000007f0000-0x000000800000 : "art"
[    0.503446] libphy: Fixed MDIO Bus: probed
[    0.867722] ag71xx 19000000.eth: Could not connect to PHY device. Deferring probe.
[    1.547071] libphy: ag71xx_mdio: probed
[    1.552183] libphy: ar8xxx-mdio: probed
[    1.563432] switch0: Atheros AR8229 rev. 1 switch registered on mdio-bus.0
[    1.615688] ag71xx 1a000000.eth: connected to PHY at fixed-0:00 [uid=00000000, driver=Generic PHY]
[    1.625625] eth0: Atheros AG71xx at 0xba000000, irq 5, mode: gmii
[    1.635048] NET: Registered protocol family 10
[    1.646594] Segment Routing with IPv6
[    1.650534] NET: Registered protocol family 17
[    1.655143] 8021q: 802.1Q VLAN Support v1.8
[    1.998615] ag71xx 19000000.eth: connected to PHY at mdio-bus.0:1f:04 [uid=004dd042, driver=Generic PHY]
[    2.009375] eth1: Atheros AG71xx at 0xb9000000, irq 4, mode: mii
[    2.025572] VFS: Mounted root (squashfs filesystem) readonly on device 31:5.
[    2.041165] Freeing unused kernel memory: 1188K
[    2.045765] This architecture does not have kernel memory protection.
[    2.052333] Run /sbin/init as init process
[    2.546637] random: fast init done
[    3.199387] init: Console is alive
[    3.203119] init: - watchdog -
[    4.418327] kmodloader: loading kernel modules from /etc/modules-boot.d/*
[    4.600927] kmodloader: done loading kernel modules from /etc/modules-boot.d/*
[    4.618854] init: - preinit -
[    5.931873] random: jshn: uninitialized urandom read (4 bytes read)
[    6.117755] random: jshn: uninitialized urandom read (4 bytes read)
[    6.256887] random: jshn: uninitialized urandom read (4 bytes read)
[    6.867018] IPv6: ADDRCONF(NETDEV_UP): eth1: link is not ready
[    8.993059] eth1: link up (100Mbps/Full duplex)
[    8.999069] IPv6: ADDRCONF(NETDEV_CHANGE): eth1: link becomes ready
[   10.207437] jffs2: notice: (477) jffs2_build_xattr_subsystem: complete building xattr subsystem, 4 of xdatum (2 unchecked, 2 orphan) and 16 of xref (2 dead, 0 orphan) found.
[   10.227247] mount_root: switching to jffs2 overlay
[   10.257353] overlayfs: upper fs does not support tmpfile.
[   10.273488] urandom-seed: Seeding with /etc/urandom.seed
[   10.491391] eth1: link down
[   10.508995] procd: - early -
[   10.512046] procd: - watchdog -
[   11.161098] procd: - watchdog -
[   11.164669] procd: - ubus -
[   11.198778] urandom_read: 5 callbacks suppressed
[   11.198788] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.219476] random: ubusd: uninitialized urandom read (4 bytes read)
[   11.227779] procd: - init -
[   12.184526] kmodloader: loading kernel modules from /etc/modules.d/*
[   12.247737] Loading modules backported from Linux version v5.3-rc4-0-gd45331b00ddb
[   12.255420] Backport generated by backports.git v5.3-rc4-1-0-g4ec72687
[   12.347984] xt_time: kernel timezone is -0000
[   12.464698] urngd: v1.0.0 started.
[   12.620410] PPP generic driver version 2.4.2
[   12.638013] NET: Registered protocol family 24
[   12.788251] random: crng init done
[   12.804331] ath: EEPROM regdomain: 0x0
[   12.804340] ath: EEPROM indicates default country code should be used
[   12.804345] ath: doing EEPROM country->regdmn map search
[   12.804362] ath: country maps to regdmn code: 0x3a
[   12.804370] ath: Country alpha2 being used: US
[   12.804374] ath: Regpair used: 0x3a
[   12.820199] ieee80211 phy0: Selected rate control algorithm 'minstrel_ht'
[   12.822174] ieee80211 phy0: Atheros AR9340 Rev:2 mem=0xb8100000, irq=12
[   12.892979] kmodloader: done loading kernel modules from /etc/modules.d/*
[   23.047101] br-lan: port 1(eth1) entered blocking state
[   23.052417] br-lan: port 1(eth1) entered disabled state
[   23.058192] device eth1 entered promiscuous mode
[   23.076245] IPv6: ADDRCONF(NETDEV_UP): br-lan: link is not ready
[   23.137492] IPv6: ADDRCONF(NETDEV_UP): eth0: link is not ready
[   24.187933] eth0: link up (1000Mbps/Full duplex)
[   24.192663] IPv6: ADDRCONF(NETDEV_CHANGE): eth0: link becomes ready
[   25.149023] eth1: link up (100Mbps/Full duplex)
[   25.153687] br-lan: port 1(eth1) entered blocking state
[   25.159037] br-lan: port 1(eth1) entered forwarding state
[   25.291618] IPv6: ADDRCONF(NETDEV_CHANGE): br-lan: link becomes ready
[   27.765225] IPv6: ADDRCONF(NETDEV_UP): wlan0: link is not ready
[   27.776586] br-lan: port 2(wlan0) entered blocking state
[   27.782193] br-lan: port 2(wlan0) entered disabled state
[   27.788102] device wlan0 entered promiscuous mode
[   27.818094] IPv6: ADDRCONF(NETDEV_CHANGE): wlan0: link becomes ready
[   27.824806] br-lan: port 2(wlan0) entered blocking state
[   27.830274] br-lan: port 2(wlan0) entered forwarding state
```

psyborg55 commented:
ok thanks for dmesg output

psyborg55 commented:
> That's also what makes it so hard to find the problem, as we do not know what's wrong.

Start with building image at commit when support was first added. If that one works bisection is a way to go.

adrianschmutzler commented:
> Start with building image at commit when support was first added. If that one works bisection is a way to go.

;-) The problem exists since the beginning.

psyborg55 commented:
i was wondering how could incomplete/untested PR for the device get into the tree. then saw Petr Štetiar was the committer so i'm not surprised at all...

adrianschmutzler commented:
Don't you have some other place for bullying?

blocktrron commented:
@citronalco the capitalization in your commit title is currently off. First letter after the prefix has to be lower case.

citronalco commented:
@blocktrron fixed.

apcameron commented:
In target/linux/ar71xx/files/arch/mips/ath79/mach-cpe510.c
I see the following comment

/* All LEDs/button except for link4 are the same for CPE and WBS series */
#define WBS510_GPIO_LED_L4 2

Does that apply to the  WBS210 v2 as well?

adrianschmutzler commented:
That's a good find.
@citronalco Does your link4 LED blink during boot?

citronalco commented:
That's really a good find: I checked, and indeed, the LED for link 4 does not work.

adrianschmutzler commented:
This is really hard to implement in DTS nicely. I'd propose the following:
Change led label from "system" to "led_link4" in ar9344_tplink_cpexxx-v1.dtsi (unrelated, but nicer). Change aliases accordingly.
Then, remove the gpios from led_link4 in v1.dtsi and add the following to the derived DTSes:
```
&led_link4 {
	gpios = <&gpio 16 GPIO_ACTIVE_LOW>;
};
```
(with appropriate gpio for cpe/wbs).

Maybe push changes first before you test, so I can have a quick look onto them.

citronalco commented:
In the meantime I verified link4 LED is working with GPIO set to 2 instead of 16.
I'll try, but be warned: I have no real experience with DTS files up to now.

adrianschmutzler commented:
@citronalco Well, then you have the chance to learn. :-)

adschm reviewed (changes requested):
GPIO adjustments required.

citronalco commented:
OK, done. And tested.

adrianschmutzler commented:
Looks good, first file seems to have wrong indent for gpio...

citronalco commented:
Fixed. Thanks.

adrianschmutzler commented:
Thanks, pulled into my staging tree (just added a comment about the GPIO stuff).

apcameron commented:
I think this code need to change from
"SupportList:\r\n"
"WBS210(TP-LINK|UN|N300-2|00000000):2.00\r\n"
"WBS210(TP-LINK|US|N300-2|55530000):2.00\r\n"
"WBS210(TP-LINK|EU|N300-2|45550000):2.00\r\n",

To
"SupportList:\r\n"
"WBS210(TP-LINK|UN|N300-2|00000000):2.0\r\n"
"WBS210(TP-LINK|US|N300-2|55530000):2.0\r\n"
"WBS210(TP-LINK|EU|N300-2|45550000):2.0\r\n",

adrianschmutzler commented:
Unless @citronalco concludes otherwise, please propose a patch for this change.

apcameron commented:
I will leave @citronalco to look at it as I do not have a WBS210 device.

citronalco commented:
I checked, it also works with the change @apcameron proposed, so this bytes can be saved.
@adrianschmutzler: Should I open another pull request for this?

adrianschmutzler commented:
@citronalco Yes, please.

citronalco commented:
Done. See #2567
